crates/storage/src/connection/transaction.rs

```diff
@@ -483,42 +483,86 @@ pub(super) fn transaction_block_hash(
 }
 
 pub(crate) mod dto {
+    use std::fmt;
+
     use fake::{Dummy, Fake, Faker};
     use pathfinder_common::*;
     use pathfinder_crypto::Felt;
-    use serde::{Deserialize, Serialize};
-    use smallvec::SmallVec;
+    use serde::{ser::SerializeSeq, Deserialize, Serialize};
 
     /// Minimally encoded Felt value.
-    #[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq, Default)]
-    #[serde(deny_unknown_fields)]
-    pub struct MinimalFelt(SmallVec<[u8; 32]>);
+    #[derive(Clone, Debug, PartialEq, Eq, Default)]
+    pub struct MinimalFelt(Felt);
+
+    impl serde::Serialize for MinimalFelt {
+        fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+        where
+            S: serde::Serializer,
+        {
+            let len = self
+                .0
+                .to_be_bytes()
+                .into_iter()
+                .skip_while(|&x| x == 0)
+                .count();
+            let mut seq = serializer.serialize_seq(Some(len))?;
+            for elem in self.0.to_be_bytes().into_iter().skip_while(|&x| x == 0) {
+                seq.serialize_element(&elem)?;
+            }
+            seq.end()
```

Comment:
Might be able to inverse this and just serialize the non-zero bytes? Haven't actually tried this though.

Separate note: you should use `as_be_bytes()` for the first one since it borrows. Currently you're cloning the array to iterate and count.

```rust
let bytes = self.0.to_be_bytes();
let zeros = bytes.iter().take_while(|x| x == 0).count();
serializer.serialize(&bytes[zeros..])
```